packages/next/shared/lib/router/router.ts

```diff
@@ -1917,7 +1916,11 @@ export default class Router implements BaseRouter {
 
     const fns = await this.pageLoader.getMiddlewareList()
     const requiresPreflight = fns.some(([middleware, isSSR]) => {
-      return getRouteMatcher(getMiddlewareRegex(middleware, !isSSR))(cleanedAs)
+      return getRouteMatcher(
+        getMiddlewareRegex(middleware, {
+          catchAll: !isSSR,
+        })
+      )(cleanedAs)
     })
 
     if (!requiresPreflight) {
```

Comment:
Can the above code/checks all be removed in favor a single flag like `process.env.__NEXT_HAS_ROOT_MIDDLEWARE` which is set via the webpack config here? https://github.com/vercel/next.js/blob/c72d32d3b5cce4fcf33cf621f0e43b899a493424/packages/next/build/webpack-config.ts#L1522

Reply:
Since this is a quite big change already I'd love to do that in a follow-up PR if that's ok for you @ijjk WDYT?

Reply:
Yeah we can do that in a follow-up 👍